README.md
[![Build Status](https://travis-ci.com/jmacdotorg/sweat.svg?branch=master)](https://travis-ci.com/jmacdotorg/sweat)
# Sweat

A chatty, distracting, and flexible workout timer.

# Examples

Run through a seven-minute workout while your computer reads aloud trivia from
Wikipedia, reports on the weather, and tells dumb jokes:

    sweat

Run through a seven-minute workout, but without any of that other stuff:

    sweat --no-entertainment

Run through a seven-minute workout with semi-randomized drills:

    sweat --shuffle

Have the program read news headlines instead of clicking around Wikipedia:

    sweat --newsapi-key=MyNewsApiKey12345

# Description

Sweat is a workout timer that helps distract you from the pain of
exercise by chatting incessantly, using your computer's text-to-speech capabilities to read news headlines, crack
strange old-man jokes, talk about the weather, and still manage to call out
workout prompts when necessary.

Sweat is optimized for the so-called [Seven-Minute Workout (7MW)](https://well.blogs.nytimes.com/2013/05/09/the-scientific-7-minute-workout/), which
leads you through twelve 30-second drills, with 10-second rests in
between. These focus on four types of exercise: aerobic, lower-body,
upper-body, and core. While it has sensible and widely accepted
defaults, you can change or expand its list of drills if you really
want, or adjust how many drills each workout entails, or the timing
involved.

Sweat features a friendly pause function, and a shuffle mode that will
randomize the drills you receive within each type while keeping the types themselves in order,
ensuring you get a varied and balanced workout.

Sweat assumes you already know how to perform the drills it calls out,
and trusts you to get through them in whatever way works best for you.
Sweat will never judge you.

Yes, Sweat is a command-line program. Get your butt off the chair once
in a while and onto the floor, fellow hackers. It's good for you.

## How Sweat is different from other timers

Sweat's major features that most other 7MW timers don't have:

### A speech-centered UI

Sweat guides you through voice alone (with a simple text transcription in its terminal window).

### Chatty entertainment while you struggle

Sweat will (unless you ask it not to) click through a randomly chosen thread of related Wikipedia articles while you work out and read aloud what it finds.
The workout takes priority, so this intentionally distracting chatter will not make you miss any exercise cues.

With a little extra configuration, you can have Sweat instead read you current news headlines from a variety of sources.

Sweat will also tell you the local weather during certain drills, and end by reading aloud the output of the \`fortune\` program (if available).

### Optional drill-shuffling

For variety's sake, you can shuffle the drills out of their standard order. While you still get three rounds of aerobic, lower-body, upper-body, and core drills in that order, Sweat will randomize the order of the three drills within each category.

### Pause between side-switching

Halfway through the side-plank drill, Sweat gives you a few seconds to adjust yourself.
(For some reason, few if any other timers seems to bother with this.)

### No-chair and no-jumping modes

A no-chair mode substitutes other drills when you find yourself in a space (e.g. a hotel room) with no suitably stable chair to exercise with.

A no-jumping mode is also available when you want to avoid stomping around on your downstairs neighbor's ceiling.

### Lots of configuration options

It's a command-line Unixish program, so of course it's far too configurable. Happily, its defaults should fit most needs...

# Installation

First, make sure you have the \`cpanm\` program on your machine. It is likely
available as "cpanminus" in your favorite package manager. (Or install it from
source, through the instructions at [http://cpanmin.us](http://cpanmin.us).)

Then, run these commands:

    cpanm --installdeps .
    perl Build.PL
    perl Build build
    perl Build install    # run under sudo to install at system-level

# Usage

Once installed, you can get a quick-reference guide like this:

    sweat --help

Or far more thorough instructions this way:

    man sweat

# Notes

This is this module's first release (or nearly so). It works for the
author's own use-cases, but it's probably buggy beyond that. Please
report issues at [the module's GitHub
site](https://github.com/jmacdotorg/sweat). Code and documentation
pull requests are very welcome!

# Author

Jason McIntosh (jmac@jmac.org)

# Copyright and licence

This software is Copyright (c) 2019 by Jason McIntosh.

This is free software, licensed under:

    The MIT (X11) License


Changes
Revision history for Perl extension Sweat

{{$NEXT}}

202101030 2021-01-03T21:36:31Z

    - Adding a choice of news articles or Wikipedia entries to the interactive mode.

202005170 2020-05-17T13:28:23Z

    - More graceful reaction to articles failing to load at the start of the workout.

202004250 2020-04-26T00:20:22Z

    - More graceful reaction to articles failing to load midway through the workout.

202002270 2020-02-28T01:54:16Z

    - Adding a simple interactive mode.

    - Better handling of unset configuration options and connectivity problems. Thanks to grai ("sudo-batman") for the bug reports.

202002100 2020-02-10T15:40:59Z

    - Adding a "refocus" option that will attempt to refocus back to the window Sweat is running in after it opens a browser window.

    - Wikipedia articles now load in the background during the workout, rather than forcing a pre-workout wait while they all finish loading.

    - Improved CPAN distribution abstract (Thanks to Toby Inkster for the suggestion)

201912110 2019-12-11T19:38:38Z

    - Changes to Wikipedia mode:
    
        - Avoiding repeating any article during a single workout
        
        - Avoiding articles with certain keywords in the title that suggest primarily non-prose content (e.g. "List of" or "Comparison of")

        - Dropping parentheticals (e.g. birth and death dates, original-language spellings, and IPA pronunciations) from Wikipedia articles' summaries, because they often do not read out loud well.

    - Fixing a bug that would ignore "no-news" in a config file.
    
    - Fixed incorrect docs and sample-config info (Thanks to Tim Vroom)

201909240 2019-09-24T20:55:34Z

    - Adding a "language" config option for setting Wikipedia localization.
    
    - Declining to follow links to Wikipedia articles with numerals in the title, in an effort to avoid boring chains of annual-statistics pages.

    - Configuring the internal MediaWiki::API object to automatically retry a few times if any response from the wiki takes more than a couple of seconds.
    
201909171 2019-09-18T03:00:57Z

    - Making use of File::Which, rather than assuming the presence of a `command` command. (Thanks to Kivanc Yazan for pointing this problem out.)

201909170 2019-09-18T01:05:29Z

    - Adding LWP::Protocol::https as a dependency.

201909090 2019-09-09T21:32:31Z

    - Replacing the LICENSE file with an MIT license, as originally intended

    - Small organizational improvements for CPAN's sake; thanks to Dan Book for the suggestions.

201909080 2019-09-09T20:26:49Z

    - original version



lib/Sweat/Drill.pm
package Sweat::Drill;

use strictures 2;
use Types::Standard qw(Bool Str);

use List::Util qw(shuffle);

use Moo;
use namespace::clean;

has 'requires_a_chair' => (
    is => 'ro',
    default => 0,
    isa => Bool,
);

has 'name' => (
    is => 'ro',
    required => 1,
    isa => Str,
);

has 'requires_jumping' => (
    is => 'ro',
    default => 0,
    isa => Bool,
);

has 'requires_side_switching' => (
    is => 'ro',
    default => 0,
    isa => Bool,
);

has 'is_used' => (
    is => 'rw',
    default => 0,
    isa => Bool,
);

1;

=head1 Sweat::Drill - Library for the `sweat` command-line program

=head1 DESCRIPTION

This library is intended for internal use by the L<sweat> command-line program,
and as such offers no publicly documented methods.

=head1 SEE ALSO

L<sweat>

=head1 AUTHOR

Jason McIntosh <jmac@jmac.org>


lib/Sweat/Group.pm
package Sweat::Group;

use warnings;
use strict;
use Types::Standard qw(ArrayRef Str);

use List::Util qw(shuffle);
use Try::Tiny;

use Sweat::Drill;

use Moo;
use namespace::clean;

has 'drills' => (
    is => 'ro',
    required => 1,
    isa => ArrayRef,
);

has 'name' => (
    is => 'ro',
    required => 1,
    isa => Str,
);

has 'sweat' => (
    is => 'ro',
    required => 1,
    weaken => 1,
);

sub BUILD {
    my ( $self, $args ) = @_;

    # We reset the drills immediately upon building so that restrictive
    # preferences (e.g. no-jumping) can get applied immediately.
    $self->reset_drills;

}

# reset_drills: Set the "is_used" bit for each drill to 0, *unless* that drill
#               is restricted according to current config. In that case, it
#               gets "is_used" set to 1.
sub reset_drills {
    my $self = shift;

    my %restriction_map = (
        no_jumping => 'requires_jumping',
        no_chair => 'requires_a_chair',
    );

    for my $drill ( @{ $self->drills } ) {
        $drill->is_used(0);
        for my $restriction ( keys %restriction_map ) {
            my $attribute = $restriction_map{$restriction};
            if ( $drill->$attribute && $self->sweat->$restriction ) {
                $drill->is_used(1);
                last;
            }
        }
    }

    unless ($self->unused_drills) {
        die "The current configuration makes every drill in the "
            . $self->name
            . " group unavailable, and we can't have that.\n";
    }
}

sub new_from_config_data {
    my ( $class, $sweat, $data ) = @_;
    my @groups;
    try {
        for my $group_data ( @$data ) {
            my @drills;
            for my $drill_data ( @{$group_data->{drills} } ) {
                my $drill = Sweat::Drill->new( $drill_data );
                push @drills, $drill;
            }
            my $group = $class->new(
                drills => \@drills,
                name => $group_data->{name},
                sweat => $sweat,
            );
            push @groups, $group;
        }
    }
    catch {
        die "Sorry... I can't parse the provided drill-group config data. $_\n";
    };
    return @groups;
}

sub unused_drills {
    my $self = shift;

    return grep { not $_->is_used } @{ $self->drills };
}

1;

=head1 Sweat::Group - Library for the `sweat` command-line program

=head1 DESCRIPTION

This library is intended for internal use by the L<sweat> command-line program,
and as such offers no publicly documented methods.

=head1 SEE ALSO

L<sweat>

=head1 AUTHOR

Jason McIntosh <jmac@jmac.org>


lib/Sweat/Article.pm
package Sweat::Article;

use warnings;
use strict;
use Moo;
use namespace::clean;
use utf8::all;

use Types::Standard qw( Str Maybe );

use Scalar::Util qw( blessed );
use HTML::Strip;
use List::Util qw( shuffle );
use MediaWiki::API;

has 'text' => (
    is => 'ro',
    required => 1,
    isa => Str,
);

has 'title' => (
    is => 'ro',
    required => 1,
    isa => Str,
);

has 'url' => (
    is => 'ro',
    required => 1,
);

our $stripper = HTML::Strip->new;
our $mw = MediaWiki::API->new;

$mw->{config}->{max_lag} = 2;
$mw->{config}->{max_lag_delay} = 1;
our $language;

our %seen_titles;

sub new_from_newsapi_article {
    my ( $class, $newsapi_article ) = @_;

    die "Expected a NewsAPI article, got $newsapi_article"
        unless blessed($newsapi_article)
               && $newsapi_article->isa( 'Web::NewsAPI::Article' );

    my $sweat_article = $class->new(
        text => ($newsapi_article->title // q{})
                . q{. }
                . ($newsapi_article->description // q{}),
        url => $newsapi_article->url,
        title => $newsapi_article->title,
    );

    return $sweat_article;
}

sub new_from_random_wikipedia_article {
    my ($class) = @_;

    my $title = _get_random_title();
    return $class->new_from_wikipedia_title($title);
}

sub new_from_linked_wikipedia_article {
    my ($class, $article) = @_;

    my $title = _get_random_title_linked_from_title($article->title);
    return $class->new_from_wikipedia_title($title);
}

sub new_from_wikipedia_title {
    my ($class, $title) = @_;

    my $summary = _get_summary_for_title($title);
    my $tries = 0;
    until ($summary || ($tries >= 3) ) {
        $tries++;
        $title = _get_random_title_linked_from_title($title);
        $summary = _get_summary_for_title($title);
    }
    unless ( $summary ) {
        $title = _get_random_title();
        $summary = _get_summary_for_title($title);
    }

    return $class->new(
        title => $title,
        text => $summary,
        url => "https://$language.wikipedia.org/wiki/$title",
    );
}

sub _get_random_title {
    my $result = $mw->api( {
        list => 'random',
        action => 'query',
        rnnamespace => 0,
    } );

    return $result->{query}->{random}->[0]->{title};
}

sub _get_summary_for_title {
    my ($title) = @_;

    my $result = $mw->api( {
        action => 'query',
        prop => 'extracts',
        exintro => undef,
        titles => $title,
    } );

    my $summary = (values(%{$result->{query}->{pages}}))[0]->{extract};

    if (defined $summary) {
        $summary = $stripper->parse( $summary );
        # Eliminate all parentheticals (birth/death dates, alternate-language
        # representations, and so on) because they don't read out loud well.
        my $found_some;
        until ( defined($found_some) && not($found_some) ) {
            $found_some = $summary =~ s/\([^\(]*?\)//g;
        }
        # Clean up redundant whitespace, and whitespace before punctuation.
        $summary =~ s/\s+([,.!?;:])/$1/g;
        $summary =~ s/ {2,}/ /g;
    }
    if ( $summary && $summary =~ /\S/ ) {
        return $summary;
    }
    else {
        return undef;
    }
}

sub _erase_parentheticals {
    my ( $summary, $opener, $closer ) = @_;
    my $found_some;
    until ( defined($found_some) && not($found_some) ) {
        $found_some = $summary =~ s/\([^\(]*?\)//g;
        warn "Found some: $found_some\n";
    }
    return $summary;
}


sub _get_random_title_linked_from_title {
    my ($title) = @_;

    my $result = $mw->api( {
        action => 'query',
        prop => 'links',
        titles => $title,
        plnamespace => 0,
        pllimit => 100,
    } );

    my $links_ref = (values(%{$result->{query}->{pages}}))[0]->{links};

    my @links = shuffle(@$links_ref);

    my $linked_title;

    until ($linked_title || (@links == 0 )) {
        if (defined $links[0]) {
            my $proposed_title = $links[0]->{title};
            # Skip:
            # * Any title we've already seen
            # * Any title with a numeral in it (to stay away from annual-
            #   statistics gravity wells)
            # * Any title with a word suggesting it's a just a list or table
            unless (
                $seen_titles{$proposed_title}
                &&
                $proposed_title =~ /\d|^list of\s|^comparison of\s|^table of\s/i
            ) {
                $linked_title = $proposed_title;
                $seen_titles{$proposed_title} = 1;
            }
        }
        shift @links;
    }

    if ($linked_title) {
        return $linked_title;
    }
    else {
        return _get_random_title();
    }
}


1;

=head1 Sweat::Article - Library for the `sweat` command-line program

=head1 DESCRIPTION

This library is intended for internal use by the L<sweat> command-line program,
and as such offers no publicly documented methods.

=head1 SEE ALSO

L<sweat>

=head1 AUTHOR

Jason McIntosh <jmac@jmac.org>


lib/Sweat.pm
package Sweat;

our $VERSION = 202004250;

use v5.10;

use warnings;
use strict;
use Types::Standard qw(Int ArrayRef Maybe Str Bool HashRef);

use List::Util qw(shuffle);
use YAML;
use File::Temp qw(tmpnam);
use Web::NewsAPI;
use MediaWiki::API;
use LWP;
use Try::Tiny;
use utf8::all;
use Term::ReadKey;
use POSIX qw(uname);
use File::Which;

use Sweat::Group;
use Sweat::Article;
use Sweat::ArticleHandler;

use Moo;
use namespace::clean;

BEGIN {
    binmode STDOUT, ":utf8";
}

sub DEMOLISH {
    \&clean_up;
}

has 'groups' => (
    is => 'lazy',
    isa => ArrayRef,
);

has 'group_config' => (
    is => 'rw',
    isa => ArrayRef,
    default => sub { _default_group_config( @_ ) },
);

has 'drills' => (
    is => 'lazy',
    isa => ArrayRef,
);

has 'drill_count' => (
    is => 'rw',
    isa => Int,
    required => 1,
    default => 12,
);

has 'drill_counter' => (
    is => 'rw',
    isa => Int,
    default => 0,
);

has 'config' => (
    is => 'ro',
    isa => Maybe[Str],
);

has 'shuffle' => (
    is => 'rw',
    default => 0,
    isa => Bool,
);

has 'entertainment' => (
    is => 'rw',
    default => 1,
    isa => Bool,
);

has 'chair' => (
    is => 'rw',
    default => 1,
    isa => Bool,
);

has 'jumping' => (
    is => 'rw',
    default => 1,
    isa => Bool,
);

has 'drill_length' => (
    is => 'rw',
    isa => Int,
    default => 30,
);

has 'rest_length' => (
    is => 'rw',
    isa => Int,
    default => 10,
);

has 'drill_prep_length' => (
    is => 'rw',
    isa => Int,
    default => 1,
);

has 'side_switch_length' => (
    is => 'rw',
    isa => Int,
    default => 4,
);

has 'speech_program' => (
    is => 'rw',
    isa => Str,
    default => sub { (uname())[0] eq 'Darwin'? 'say' : 'espeak' },
);

has 'url_program' => (
    is => 'rw',
    isa => Maybe[Str],
    default => sub { (uname())[0] eq 'Darwin'? 'open' : undef },
);

has 'fortune_program' => (
    is => 'rw',
    isa => Maybe[Str],
    default => 'fortune',
);

has 'refocus_program' => (
    is => 'lazy',
    isa => Maybe[Str],
);

has 'refocus' => (
    is => 'rw',
    isa => Bool,
    default => 0,
);

has 'newsapi_key' => (
    is => 'rw',
    isa => Maybe[Str],
);

has 'country' => (
    is => 'rw',
    isa => Str,
    default => 'us',
);

has 'language' => (
    is => 'rw',
    isa => Str,
    default => 'en',
);

has 'article_handler' => (
    is => 'ro',
    isa => sub { $_[0]->isa('Sweat::ArticleHandler') },
    default => sub { Sweat::ArticleHandler->new },
    handles => [ qw( add_article next_article ) ],
);

has 'weather' => (
    is => 'lazy',
    isa => Str,
);

has 'speaker_pid' => (
    is => 'rw',
    isa => Int,
);

around BUILDARGS => sub {
    my ( $orig, $class, %args ) = @_;

    _mangle_args( \%args );

    return $class->$orig(%args);

};

sub BUILD {
    my ($self, $args) = @_;

    my $config;

    try {
        $config = Load( $self->config );
    }
    catch {
        die "Can't load the config file! Here's the error from the YAML parser: "
            . $_
            . "\n";
    };

    foreach ( $args, $config ) {
        _mangle_args( $_ );
    }

    for my $method( qw(shuffle entertainment chair jumping refocus)) {
        next if defined $args->{$method};
        if ( defined $config->{$method} ) {
            my $value = $config->{$method} // 0;
            $self->$method($value);
        }
    }

    for my $method (
        qw(
            newsapi_key country fortune_program speech_program url_program
            rest_length drill_length drill_count
        )
    ) {
        next if defined $args->{$method};
        if ( defined $config->{$method} ) {
            $self->$method($config->{$method});
        }
    }

    if ( my $group_data = $config->{groups} ) {
        $self->group_config( $group_data );
    }

    if ( $config->{no_news} ) {
        $self->newsapi_key( undef );
    }

    $Sweat::Article::language = $self->language;
    $Sweat::Article::mw->{config}->{api_url} =
        "https://"
        . $self->language
        . ".wikipedia.org/w/api.php";

    $self->_check_resources;
}

sub _check_resources {
    my $self = shift;

    my $speech_program = $self->speech_program;
    my $bare_speech_program = (split /\s+/, $speech_program)[0];

    unless ( which( $bare_speech_program) ) {
        die "ERROR: Sweat's 'speech-program' configuration is set to "
            . "'$speech_program', but there doesn't seem to be a program "
            . "there. I can't run without a speech program... sorry!\n";
    }

    if ( $self->entertainment ) {
        foreach (qw (url fortune) ) {
            my $method = "${_}_program";
            my $program = $self->$method;
            next unless defined $program;
            my $bare_program = (split /\s+/, $program)[0];
            unless ( which( $bare_program ) ) {
                $self->$method( undef );
                warn "WARNING: Sweat's '$_-program' configuration is set to "
                     . "'$program', but there doesn't seem to be a program there. "
                     . "Going ahead without $_-opening.\n";
            }
        }
    }
}


sub _load_entertainment {
    my $self = shift;
    if ( $self->entertainment ) {
        local $| = 1;
        say "Loading entertainment, please wait...";
        $self->_load_articles;
        $self->weather;
        say "...done.";
    }

}

sub _load_articles {
    my $self = shift;

    try {
        if ( $self->newsapi_key ) {
            my $newsapi = Web::NewsAPI->new(
                api_key => $self->newsapi_key,
            );
            my $result = $newsapi->top_headlines(
                country => $self->country,
                pageSize => $self->drill_count,
            );
            foreach ( $result->articles ) {
                $self->add_article(
                    Sweat::Article->new_from_newsapi_article($_)
                );
            }
        }
        else {
            my $article = Sweat::Article->new_from_random_wikipedia_article;
            unless ( fork ) {
                $self->add_article( $article );
                for (1..$self->drill_count) {
                    $article = Sweat::Article->
                               new_from_linked_wikipedia_article($article);
                    $self->add_article( $article );
                }
                exit;
            }
        }
    }
    catch {
        clean_up();
        die "Sweat ran into a problem fetching articles. Check your internet\n"
            . "connection, or run with --no-entertainment to use sweat without\n"
            . "loading any online distractions.\n\n(Error text: $_)\n";
    }
}

my $temp_file = tmpnam();

sub sweat {
    my $self = shift;

    ReadMode 3;

    $self->_load_entertainment;

    for my $drill (@{ $self->drills }) {
        $self->order($drill);
    }

    $self->cool_down;
    $self->clean_up;
}

sub order {
    my ( $self, $drill ) = @_;

    $self->drill_counter( $self->drill_counter + 1 );

    $self->rudely_speak(
        'Prepare for '
        . $drill->name
        . '. Drill '
        . $self->drill_counter
        . '.'
    );
    $self->countdown($self->rest_length);

    my ($extra_text, $url, $article) = $self->entertainment_for_drill( $drill );
    $extra_text //= q{};

    # XXX Set the binmode to raw temporarily if we're side-switching (and thus
    #     reporting weather.) This is a hack to get around a strange encoding
    #     issue in this case only.
    binmode STDOUT, ':raw' if $drill->requires_side_switching;
    $self->rudely_speak( "Start now. $extra_text");
    binmode STDOUT, ':utf8' if $drill->requires_side_switching;

    my $url_tempfile;
    if ( defined $url ) {
        if ( $url =~ m{\Wyoutube.com/} ) {
            $url_tempfile = $self->mangle_youtube_url( $article );
            $url = "file://$url_tempfile";
        }
        if ( defined $self->url_program ) {
            system( split (/\s+/, $self->url_program), $url );
            if ( defined $self->refocus_program ) {
                system $self->refocus_program;
            }
        }
    }

    if ( $drill->requires_side_switching ) {
        $self->countdown( $self->drill_length / 2 );
        $self->speak( 'Switch sides.');
        $self->countdown( $self->side_switch_length );
        $self->speak( 'Resume.' );
        $self->countdown( $self->drill_length / 2 );
    }
    else {
        $self->countdown( $self->drill_length );
    }

    $self->rudely_speak( 'Rest.' );
    sleep $self->drill_prep_length;
}

sub countdown {
    my ($self, $seconds) = @_;
    return unless $seconds;

    my $seconds_label_cutoff = 10;
    my @spoken_seconds = (20, 15, 10, 5, 4, 3, 2, 1);
    my $label = 'seconds left';

    for my $current_second (reverse(0..$seconds)) {
        my $keystroke = ReadKey (1);
        if ( $keystroke ) {
            $self->pause;
        }
        if (
            ( grep {$_ == $current_second} @spoken_seconds )
            || ( $current_second && not ( $current_second % 10 ) )
        ) {
            if ( $current_second >= $seconds_label_cutoff ) {
                $self->speak( "$current_second $label." );
                $label = 'seconds';
            }
            else {
                $self->shut_up; # Final countdown, so interrupt any chattiness
                $self->speak( $current_second );
            }
        }
    }
}

sub shut_up {
    my $self = shift;

    # We check for drill-length here for the sake of testing, which usually
    # sets this to 0, and also runs under the `prove` program... which will
    # get killed by the code below, oops. Something to improve later.
    if ( -e $temp_file && $self->drill_length ) {
        my $group = getpgrp;
        unlink $temp_file;
        $SIG{TERM} = 'IGNORE';
        kill ('TERM', -$group);
        $SIG{TERM} = 'DEFAULT';
    }
}

sub pause {
    my $self = shift;

    $self->shut_up;
    say "***PAUSED*** Press any key to resume.";
    $self->leisurely_speak( 'Paused.' );
    ReadKey (0);
    say "Resuming...";
    $self->leisurely_speak( 'Resuming.' );
}

sub entertainment_for_drill {
    my ( $self, $drill ) = @_;

    unless ($self->entertainment) {
        return (undef, undef);
    }

    my $text;
    my $url;
    my $article;

    if ( $drill->requires_side_switching ) {
        $text = $self->weather;
    }
    else {
        $article = $self->next_article;
        if ( $article ) {
            $text = $article->text;
            $url = $article->url;
        }
        else {
            $text = "Couldn't fetch article, so have this instead: "
                    . $self->fortune;
        }
    }

    return ( $text, $url, $article );
}


sub fortune {
    my $self = shift;

    return q{} unless $self->entertainment;

    my $exec = $self->fortune_program;

    no warnings;
    my $text = `$exec` // '';
    use warnings;

    unless (length $text) {
        warn "Sweat tried to fetch a fortune by running `$exec`, but didn't "
             . "get any output. Sorry!\n";
    }

    return $text;
}

sub _build_weather {
    my $self = shift;

    my $weather;
    my $ua = LWP::UserAgent->new;
    my $response = $ua->get('http://wttr.in?format=3');
    if ( $response->is_success ) {
        $weather = 'The current weather in ' . $response->content;
    }
    else {
        $weather = 'Could not fetch the current weather. Sorry about that.';
    }
    return $weather;
}

sub cool_down {
    my $self = shift;

    $self->speak(
        'Workout complete. ' . $self->fortune
    );
}

sub speak {
    my ( $self, $message ) = @_;

    say $message;

    return if -e $temp_file;

    my $pid = fork;
    if ( $pid ) {
        $self->speaker_pid( $pid );
    }
    else {
        open my $fh, '>', $temp_file;
        print $fh $pid;
        system ( split (/\s+/, $self->speech_program), $message );
        unlink $temp_file;
        exit;
    }
}

sub leisurely_speak {
    my ( $self, $message ) = @_;

    system ( split (/\s+/, $self->speech_program), $message );
}

sub rudely_speak {
    my ( $self, $message ) = @_;

    $self->shut_up;
    $self->speak( $message );
}

# mangle_youtube_url: create a local file that just embeds the youtube
#                     video, preventing autoplay.
sub mangle_youtube_url {
    my ( $self, $article ) = @_;
    my $tempfile = File::Temp->new( SUFFIX => '.html' );
    binmode $tempfile, ":utf8";
    my $title = $article->title;
    my $description = $article->text;
    my $url = $article->url;

    my ($video_id) = $url =~ m{v=(\w+)};

    print $tempfile qq{<html><head><title>$title</title></head>}
              . qq{<body><h1>$title</h1>}
              . qq{<p>$description</p>}
              . qq{<div><iframe src="https://www.youtube.com/embed/$video_id" }
              . q{frameborder="0" allow="accelerometer; autoplay; }
              . q{encrypted-media; gyroscope; picture-in-picture" }
              . q{allowfullscreen></iframe></div>}
              . q{<p><em>This article was mangled by Sweat so that its video }
              . q{didn&#8217;t autoplay. You&#8217;re welcome.</em></p>}
              . qq{</body></html>\n};

    return $tempfile;
}

sub _build_drills {
    my $self = shift;

    my @final_drills;

    while (@final_drills < $self->drill_count) {
        for my $group ( @{ $self->groups } ) {
            my @drills = $group->unused_drills;

            unless ( @drills ) {
                $group->reset_drills;
                @drills = $group->unused_drills;
            }

            if ( $self->shuffle ) {
                @drills = shuffle(@drills);
            }

            $drills[0]->is_used( 1 );
            push @final_drills, $drills[0];

            last if @final_drills == $self->drill_count;
        }
    }

    return \@final_drills;
}

sub clean_up {
    ReadMode 0;
}

sub _build_groups {
    my $self = shift;

    return [ Sweat::Group->new_from_config_data( $self, $self->group_config ) ];

}

sub _default_group_config {
    my $self = shift;
    return Load(<<END);
- name: aerobic
  drills:
    - name: jumping jacks
      requires_jumping: 1
    - name: high knees
      requires_jumping: 1
    - name: step-ups
      requires_a_chair: 1
- name: lower-body
  drills:
    - name: wall sit
    - name: squats
    - name: knee lunges
- name: upper-body
  drills:
    - name: push-ups
    - name: tricep dips
      requires_a_chair: 1
    - name: rotational push-ups
- name: core
  drills:
    - name: abdominal crunches
    - name: plank
    - name: side plank
      requires_side_switching: 1
END
}

sub _mangle_args {
    my ( $args ) = @_;

    for my $key (keys %$args) {
        if ( $key =~ /-/ ) {
            my $new_key = $key;
            my $value = $$args{$key};
            $new_key =~ s/-/_/g;
            $$args{$new_key} = $value;
            delete $$args{$key};
        }
    }
}

sub _build_refocus_program {
    my $self = shift;

    return unless $self->refocus;

    my $pstree = `pstree -p $$`;

    if ( (uname())[0] eq 'Darwin') {
        unless ( which( 'pstree' ) ) {
            warn "The refocus feature requires the 'pstree' program to be "
                . "installed. I can't find it, so I won't be able to "
                . "refocus the window during the workout. Sorry...\n";
            return;
        }

        my ($app) = $pstree =~ /(\S+Applications\S+)/;
        if ($app) {
            return "open -a $app";
        }
        else {
            warn "I can't figure out what terminal I'm running in?! "
                 . "I won't be able to refocus the window during the "
                 . "workout. Sorry...\n";
            return;
        }
    }
    else {
        unless ( which( 'xdotool' ) ) {
            warn "The refocus feature requires the 'xdotool' program to be "
                . "installed. I can't find it, so I won't be able to "
                . "refocus the window during the workout. Sorry...\n";
            return;
        }

        my $window_id = `xdotool getactivewindow`;
        if ( $window_id ) {
            return "xdotool windowactivate $window_id";
        }
        else {
            warn "I can't figure out what window I'm running in?! "
                 . "I won't be able to refocus the window during the "
                 . "workout. Sorry...\n";
            return;
        }
    }

}

1;

=head1 NAME

Sweat - Library for the `sweat` command-line program

=head1 DESCRIPTION

This library is intended for internal use by the L<sweat> command-line program,
and as such offers no publicly documented methods.

=head1 SEE ALSO

L<sweat>

=head1 AUTHOR

Jason McIntosh <jmac@jmac.org>
